src/telemetry/opentelemetry.ts

```diff
@@ -0,0 +1,233 @@
+import type { Message } from '@google-cloud/pubsub';
+
+import { HttpInstrumentation } from '@opentelemetry/instrumentation-http';
+import { IORedisInstrumentation } from '@opentelemetry/instrumentation-ioredis';
+import { PgInstrumentation } from '@opentelemetry/instrumentation-pg';
+import { FastifyInstrumentation } from '@opentelemetry/instrumentation-fastify';
+import { GraphQLInstrumentation } from '@opentelemetry/instrumentation-graphql';
+import { PinoInstrumentation } from '@opentelemetry/instrumentation-pino';
+import { GrpcInstrumentation } from '@opentelemetry/instrumentation-grpc';
+import { TypeormInstrumentation } from 'opentelemetry-instrumentation-typeorm';
+
+import {
+  NodeSDK,
+  logs,
+  node,
+  api,
+  resources,
+  metrics,
+} from '@opentelemetry/sdk-node';
+import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
+import { OTLPTraceExporter } from '@opentelemetry/exporter-trace-otlp-grpc';
+import { SemanticAttributes as SemAttr } from '@opentelemetry/semantic-conventions';
+import { TraceExporter } from '@google-cloud/opentelemetry-cloud-trace-exporter';
+import { MetricExporter } from '@google-cloud/opentelemetry-cloud-monitoring-exporter';
+import { CloudPropagator } from '@google-cloud/opentelemetry-cloud-trace-propagator';
+
+import { containerDetector } from '@opentelemetry/resource-detector-container';
+import { gcpDetector } from '@opentelemetry/resource-detector-gcp';
+
+import dc from 'node:diagnostics_channel';
+const channel = dc.channel('fastify.initialization');
+
+const isProd = process.env.NODE_ENV === 'production';
+const isTest = process.env.NODE_ENV === 'test';
+const resourceDetectors = [
+  resources.envDetectorSync,
+  resources.hostDetectorSync,
+  resources.osDetectorSync,
+  resources.processDetectorSync,
+  containerDetector,
+  gcpDetector,
+];
+
+const traceExporter = isProd
+  ? new TraceExporter()
+  : new OTLPTraceExporter({
+      // hostname: 'jaeger-collector',
+      url: `http://${process.env.OTLP_COLLECTOR_HOST}:${process.env.OTLP_COLLECTOR_PORT}`,
+    });
+
+const spanProcessor = isProd
+  ? new node.BatchSpanProcessor(traceExporter)
+  : new node.SimpleSpanProcessor(traceExporter);
+
+const metricReader = isProd
+  ? new metrics.PeriodicExportingMetricReader({
+      exportIntervalMillis: 10_000,
+      exporter: new MetricExporter(),
+    })
+  : new PrometheusExporter({}, () => {
+      const { endpoint, port } = PrometheusExporter.DEFAULT_OPTIONS;
+      console.log(`metrics endpoint: http://localhost:${port}${endpoint}`);
+    });
+
+// Try to get the app version from the header, then query param, then default to unknown
+export const getAppVersion = (req: FastifyRequest): string => {
+  return req.headers['x-app-version'] || req.query['v'] || 'unknown';
+};
+
+export const SemanticAttributes = {
+  ...SemAttr,
+  DAILY_APPS_VERSION: 'dailydev.apps.version',
+  DAILY_APPS_USER_ID: 'dailydev.apps.userId',
+  DAILY_HEIMDALL_SESSION: 'dailydev.heimdall.session',
```

Comment:
Why do we want to log heimdall session? From security perspective I don't feel very comfortable with that

Reply:
I'll remove that one. Just added some examples of things that could be added to traces